tensorly/regression/cp_plsr.py

```diff
@@ -0,0 +1,255 @@
+            )  # Y -= T pinv(T) u q'
+
+        return self
+
+    def predict(self, X):
+        """Returns the predicted labels for a new data tensor
+
+        Parameters
+        ----------
+        X : ndarray
+            tensor data of shape (n_samples, N1, ..., NS)
+        """
+        if self.X_shape[1:] != T.shape(X)[1:]:
+            raise ValueError(
+                f"Training X has shape {self.X_shape}, while the new X has shape {T.shape(X)}"
+            )
+        X -= self.X_mean
+        factors_kr = khatri_rao(self.X_factors, skip_matrix=0)
+        unfolded = unfold(X, 0)
+        scores = T.lstsq(factors_kr, T.transpose(unfolded))[0]  # = Tnew
+        estimators = T.lstsq(self.X_factors[0], self.Y_factors[0])[0]
+        return (
+            T.dot(
+                T.dot(T.transpose(scores), estimators), T.transpose(self.Y_factors[1])
+            )
+            + self.Y_mean
+        )
+
+    def transform(self, X, Y=None):
+        """Apply the dimension reduction.
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Samples to transform.
+        Y : array-like of shape (n_samples, n_targets), default=None
+            Target vectors.
+        Returns
+        -------
+        X_scores, Y_scores : array-like or tuple of array-like
+            Return `X_scores` if `Y` is not given, `(X_scores, Y_scores)` otherwise.
+        """
+        if self.X_shape[1:] != T.shape(X)[1:]:
+            raise ValueError(
+                f"Training X has shape {self.X_shape}, while the new X has shape {T.shape(X)}"
+            )
+        X = T.copy(X)
+        X -= self.X_mean
+        X_scores = T.zeros((T.shape(X)[0], self.n_components))
+
+        for a in range(self.n_components):
+            X_scores = T.index_update(
+                X_scores,
+                T.index[:, a],
+                multi_mode_dot(
+                    X, [ff[:, a] for ff in self.X_factors[1:]], range(1, T.ndim(X))
+                ),
+            )
+            X -= CPTensor(
+                (
+                    None,
+                    [T.reshape(X_scores[:, a], (-1, 1))]
+                    + [T.reshape(ff[:, a], (-1, 1)) for ff in self.X_factors[1:]],
+                )
+            ).to_tensor()
+
+        if Y is not None:
+            Y = T.copy(Y)
+            # Check on the shape of Y
+            if (T.ndim(Y) != 1) and (T.ndim(Y) != 2):
+                raise ValueError("Only a matrix (2-mode tensor) Y is allowed.")
+            if T.ndim(Y) == 1:
+                Y = T.reshape(Y, (-1, 1))
+            if self.Y_shape[1:] != T.shape(Y)[1:]:
+                raise ValueError(
+                    f"Training Y has shape {self.Y_shape}, while the new Y has shape {T.shape(Y)}"
+                )
+
+            Y -= self.Y_mean
+            Y_scores = T.zeros((T.shape(Y)[0], self.n_components))
+            for a in range(self.n_components):
```

Comment:
@cyrillustan why does Y_scores have to be solved component-by-component if it is at most 2D? Shouldn't it just be a simple `Y_scores = T.dot(Y, Y_factors)` like in PLS?

Reply:
Sure there is probably a way to do it without solving by component. I just simply copied this code from above.

Reply:
Turns out that solving all columns together wouldn't give you the same answer. Plugging in the original Y into `transform()`, neither way of solving can recover the Y score at fitting perfectly (just like in regular PLSR), but the one solved component by component is closer.